code/parse/sexp_container.h

```diff
@@ -28,18 +28,30 @@ enum class ContainerType : int {
 	SAVE_TO_PLAYER_FILE = 0x40000000,
 };
 
-inline constexpr bool operator&(ContainerType lhs, ContainerType rhs) {
+inline constexpr bool operator&(ContainerType lhs, ContainerType rhs)
+{
 	using ContainerTypeInt = std::underlying_type<ContainerType>::type;
 	return static_cast<ContainerTypeInt>(lhs) & static_cast<ContainerTypeInt>(rhs);
 }
 
-inline constexpr ContainerType operator|(ContainerType lhs, ContainerType rhs) {
+// equivalent of ct &= ~ct_to_remove
+inline void remove_container_type(ContainerType &ct, ContainerType ct_to_remove)
+{
+	using ContainerTypeInt = std::underlying_type<ContainerType>::type;
+	ct = static_cast<ContainerType>(static_cast<ContainerTypeInt>(ct) & ~static_cast<ContainerTypeInt>(ct_to_remove));
+}
+
+inline constexpr ContainerType operator|(ContainerType lhs, ContainerType rhs)
+{
 	using ContainerTypeInt = std::underlying_type<ContainerType>::type;
 	return static_cast<ContainerType>(static_cast<ContainerTypeInt>(lhs) |
 		static_cast<ContainerTypeInt>(rhs));
 }
 
-inline void operator|=(ContainerType lhs, ContainerType rhs) { lhs = lhs | rhs; }
+inline void operator|=(ContainerType &lhs, ContainerType rhs)
+{
+	lhs = lhs | rhs;
+}
```

Comment:
I'm pretty sure compound assignment operators need to return a reference to themselves.  The return type should be `ContainerType&` instead of `void`, and the function should return `lhs`.

Reply:
FWIW based on the fact that the code changes for PR #3463 compile, it seems like a `void` return type could work. But even if that's true, following conventions is a good thing, and maybe being able to chain `|=` will come in handy, anyway.

Reply:
So [cppreference says](https://en.cppreference.com/w/cpp/language/operators) says
>Besides the restrictions above, the language puts no other constraints on what the overloaded operators do, or on the return type (it does not participate in overload resolution), but in general, overloaded operators are expected to behave as similar as possible to the built-in operators

And further down, it says bitwise operators are expected (not not required) to follow the pattern of arithmetic operators, and the arithmetic operators like `+=`, `-=`, etc. all return a reference.

But even more importantly, that page also says
>User-defined classes and enumerations that implement the requirements of [BitmaskType](https://en.cppreference.com/w/cpp/named_req/BitmaskType) are required to overload the bitwise arithmetic operators `operator&`, `operator|`, `operator^`, `operator~`, `operator&=`, `operator|=`, and `operator^=`

And once I do all of those overloads, we can get rid of the super-ugly `remove_container_type()` and then use the natural syntax like
```
container.type &= ~ContainerType::STRING_DATA;
```
or even
```
container.type &= ~ContainerType::STRING_DATA |= ContainerType::N_DATA;
```
And from the [BitMaskType](https://en.cppreference.com/w/cpp/named_req/BitmaskType) page , it sounds like the fact that, for example, `~ContainerType::STRING_DATA` is not a meaningful container type is irrelevant.

So looks like I have some more work to do.

Reply:
That's a good bit of research - which I missed when checking the compound assignment return type yesterday - and should certainly enhance the API usability.